### PR TITLE
server: Support explicitly getting health feedback information by calling RPC

### LINKED ISSUE #16297: Performance/health feedback to client and the Unified Health Controller
## Development Task

Currently, there are several ways to evaluate whether a TiKV node is in healthy state. They are:

* SlowScore which is a way to evaluate whether raftstore is in a abnormally slow state. It can be informed to PD which enables PD to trigger the evict-slow-store-scheduler to avoid write requests to be sent to the store.
* SlowTrend, which is improved from SlowScore
* gRPC HealthService, which is a way built-in in gRPC for indicating whether the service is available to serve.

And all these things are managed in PdWorker.

We found the evict-slow-store-scheduler mentioned before is quite helpful to solve the problem that some TiKV nodes encountering slow IO might significantly affect the whole cluster's performance. However, there can still be problem if follower read is being used in the cluster, since it doesn't need to be processed on leader.

To solve the problem, we want the client-go to know whether each TiKV node is abnormal and avoid sending follower read requests to the problematic ones. We are now considering making it able to send some of the health information to the client via kv responses, so that the client has a more efficient and up-to-time information about the TiKV nodes' status and adjust the policy to select replicas.

We are planning to add a component named Unified Health Controller, which will be the unified entrance for managing and accessing the health status of the TiKV node. The SlowScore, SlowTrend and gRPC HealthService mentioned above should be moved to it, and PdWorker will still be responsible for updating them. The component itself should be outside PdWorker, which enables us to access it elsewhere, or add more information to it that are not proper to be updated in PdWorker (e.g., readpool's stats).

### TiKV

- [x] Update kvproto to support sending feedback information
    - https://github.com/pingcap/kvproto/pull/1222
- [x] Add HealthController to TiKV, move SlowScore and, SlowTrend to it
    - https://github.com/tikv/tikv/pull/16456
- [x] Send SlowScore as the feedback information to client-go
    - https://github.com/tikv/tikv/pull/16498

### PD

- ~PDClient support getting `StoreStats` field in GetStore. It's already included by the RPC protocol but not returned by the golang PDClient.~

### client-go

- [x] Receive the feedback information and maintain it as part of the store info. 
    - https://github.com/tikv/client-go/pull/1153
- [x] New replica selection implementation that takes the feedback information into consideration.
    - https://github.com/tikv/client-go/pull/1142
- [x] When a request timed out, reload region info quicklier (instead of retrying 10 times as is currently implemented) as the target TiKV might be alive but not serveable.
  - Ref: https://github.com/pingcap/tidb/issues/50432
- [ ] Support actively requesting health feedback form tikv when it's not receiving it for too long (perhaps due to requests are all evicted from that store so that the connection is idle).

Other related works in client-go repo: https://github.com/tikv/client-go/issues/1104

### TiDB

Dependencies of TiDB repo needs to be updated several times, but there isn't any major development task in TiDB repo.
Ref: https://github.com/pingcap/tidb/issues/51412

- [x] https://github.com/pingcap/tidb/pull/51413

### Next step

* Including read path status (e.g. read pool busy)
* TBD.

### PR DESCRIPTION
**Pull request by MyonKeminta**
### What is changed and how it works?

Issue Number: Close #16297

What's Changed:

```commit-message
This PR makes TiKV support explicitly getting health feedback information by calling RPC. Both non-batched mode and batched mode (using BatchCommands stream) are supported.
There's some special behavior when used in batched RPC mode: The BatchCommandsResponse that contains the response of getting health feedback will always have feedback information attached (in the same way as how it's attached without being explicitly requested), and the attached information and the information carried in each single GetHealthFeedbackResponse-s is the same.
```

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

- Requires: https://github.com/pingcap/kvproto/pull/1239

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by MyonKeminta** [comment]
cc @crazycs520  PTAL thanks!

File: src/server/service/kv.rs [hunk 1]
```diff
@@ -2576,6 +2612,24 @@ impl HealthFeedbackAttacher {
     }
 
     fn attach_if_needed(&mut self, resp: &mut BatchCommandsResponse) {
+        let mut requested_feedback = None;
+
+        for r in resp.responses.iter_mut().map(|r| &mut r.cmd).flatten() {
+            if let BatchCommandsResponse_Response_oneof_cmd::GetHealthFeedback(r) = r {
+                if requested_feedback.is_none() {
+                    requested_feedback = Some(self.gen_health_feedback_pb());
+                }
```
Comment (crazycs520) [inline src/server/service/kv.rs:2621]: When will this logic be used?

File: src/server/service/kv.rs [hunk 2]
```diff
@@ -1042,7 +1042,9 @@ impl<E: Engine, L: LockManager, F: KvFormat> Tikv for Service<E, L, F> {
             GRPC_RESP_BATCH_COMMANDS_SIZE.observe(r.request_ids.len() as f64);
             // TODO: per thread load is more reasonable for batching.
             r.set_transport_layer_load(grpc_thread_load.total_load() as u64);
+
```
Comment (cfzjywxk) [inline src/server/service/kv.rs:1045]: Unrelated blank lines?

File: src/server/service/kv.rs (hunk 1 above)
Comment (cfzjywxk) [reply src/server/service/kv.rs:2621]: There is explicit health fetch request in the batch commands I think.

**Review by cfzjywxk — APPROVED** [review]
LGTM

File: src/server/service/kv.rs (hunk 1 above)
Comment (MyonKeminta) [reply src/server/service/kv.rs:2621]: To make it a complete API, `GetHealthFeedbackResponse` carries the health feedback information. We don't actually use it for now, but it makes further potential uses possible. When it's used in batch_commands, the feedback information will also be attached to the `BatchCommandsResponse`, so that it can be easily handled by the batch client, and it's transparent to the batch client whether the feedback is periodically sent or explicitly requested.
Here the code is used to check a `BatchCommandsResponse`, and attach the information if explicitly requested (by checking whether there is one or more `GetHealthFeedbackResponse`).

**Review by crazycs520 — APPROVED** [review]

**Review by crazycs520 — APPROVED** [review]

**Review by ekexium — APPROVED** [review]
